grug/app/src/app.rs

```diff
@@ -294,43 +294,37 @@ where
     // 2. `authenticate`, where the sender account authenticates the transaction.
     pub fn do_check_tx(&self, tx: Tx) -> AppResult<Outcome> {
         let buffer = Shared::new(Buffer::new(self.db.state_storage(None)?, None));
-        let block = LAST_FINALIZED_BLOCK.load(&buffer)?;
-        let gas_tracker = GasTracker::new_limited(tx.gas_limit);
         let mut events = vec![];
 
-        match do_withhold_fee(
-            self.vm.clone(),
-            Box::new(buffer.clone()),
+        let mut ctx = AppCtx::new(
+            LAST_FINALIZED_BLOCK.load(&buffer)?,
+            CHAIN_ID.load(&buffer)?,
             GasTracker::new_limitless(),
-            block,
-            &tx,
-            AuthMode::Check,
-        ) {
+            Box::new(buffer) as Box<dyn Storage>,
+            self.vm.clone(),
+        );
+
+        match do_withhold_fee(ctx.clone(), &tx, AuthMode::Check) {
             Ok(new_events) => {
                 events.extend(new_events);
             },
             Err(err) => {
-                return Ok(new_outcome(gas_tracker, Err(err)));
+                return Ok(new_outcome(ctx.gas_tracker, Err(err)));
             },
         }
 
-        match do_authenticate(
-            self.vm.clone(),
-            Box::new(buffer),
-            gas_tracker.clone(),
-            block,
-            &tx,
-            AuthMode::Check,
-        ) {
+        ctx.gas_tracker = GasTracker::new_limited(tx.gas_limit);
```

Comment:
I don't like replacing a field in the context. It feels unsafe, not a good pattern.